### ISSUE TITLE
BUG: 49.7 Thunderbolt 20 does full damage on a glancing blow.

### ISSUE THREAD

**Issue by MarcJC** [MegaMek/megamek#3548]
[Tbolt Issue.zip](https://github.com/MegaMek/megamek/files/8405382/Tbolt.Issue.zip)

Version 49.7 
Attached is save game, logs, and custom folder.  

I was testing a custom King Crab with Thunderbolt 20s and noticed that when getting a glancing blow on a target it did full damage of 20 points to the target.  There didn't seem to be an armor weakness or anything as the other T-Bolt 20 hit the same target with a regular hit and also did 20 dmg.  I think a second mech had a T-bolt 20 glance in the same round later and again it did the full 20 points of damage.  

My guess is that since the weapon is a missile type the game expects it to use the cluster table and is attempting to apply a -4 malus to the cluster roll since it glanced, but since T-bolts are single missiles it just hits with 1 missile and does full damage as missiles are expected to do after the cluster roll.    

Please let me know if there is any thing else I can do to help with this issue.  

Snippet of round report:  

Weapons fire for King Crab KGC-0020T (Marc)
     Thunderbolt 20 at Dominator B-UK (Princess Center);  needs 7, rolls 9 : 
        Dominator B-UK (Princess Center) takes 20 damage to TU.
            26 Armor remaining.

     Thunderbolt 20 at Dominator B-UK (Princess Center);  needs 7, rolls 7 :  - Glancing Blow -
        Dominator B-UK (Princess Center) takes 20 damage to RS.
        Needs 7+ to destroy Searchlight, rolls 6.
            36 Armor remaining.

**Comment by HammerGS** [comment]
Save game with Glancing blows and T-Bolts set up. Save at start of firing phase. Avatars vs Lancelots. Testing show something is off. 

[Glancing.sav.gz](https://github.com/MegaMek/megamek/files/8405620/Glancing.sav.gz)

**Comment by MarcJC** [comment]
As found later in the game a T-Bolt hitting a mech with narrow/low profile quirk and only beating the target number by 1 still does full damage to the target even though the quirk should half the damage.

**Comment by Saklad5** [comment]
This issue is present in 0.49.19 and 0.49.19.1 as well.

**Comment by Saklad5** [comment]
Other missile weapons that don't roll on the cluster hits table exhibit incorrect behavior for Glancing/Direct Blows as well. A refactor of the relevant code in `MissileWeaponHandler` would probably be the best way to address this.
<img width="847" alt="image" src="https://github.com/MegaMek/megamek/assets/4258823/fb093252-eaf8-42a6-ba60-40a997d1938d">
<img width="835" alt="image" src="https://github.com/MegaMek/megamek/assets/4258823/f39b21de-dab3-4d6a-9f66-42554ee48af1">

**Comment by Saklad5** [comment]
I think the best approach would be to alter the behavior in `MissileWeaponHandler` is the cluster size is, in fact, 1.

**Comment by MarcJC** [comment]
This may be something for the official boards but if say a battle armor has a SRM 1 launcher and gets a glancing blow how does that get handled?  

Obviously anything rolling a cluster of 1 seems like it should be subject to the glancing blow damage divisor rather than the cluster modifier.  But that may not be the official ruling.  

Its also been a while since I played but how does a RAC 5 firing multiple shots work with a glancing blow.  It could modify the cluster roll by -2 or half each 5 point hit.  In fact maybe it should do both?  I really don't remember how strange things like these combine with the glancing blow rules.    

Also should an LRM 1 fired from a battle armor suit that gets a glancing blow be unaffected?  

Late night musings.  They may or may not help.

    On Wednesday, July 10, 2024 at 09:36:54 PM EDT, Jeremy Saklad ***@***.***> wrote:  
 
 


I think the best approach would be to alter the behavior in MissileWeaponHandler is the cluster size is, in fact, 1.

**Comment by Sleet01** [comment]
Another DB/GB/NLP interaction that needs updating:

![image](https://github.com/user-attachments/assets/7f4487aa-09a7-441f-9e57-f4e8d45dd6bd)

See:
https://bg.battletech.com/forums/index.php?topic=82273.msg1960654#msg1960654

I'd guess this needs a specific handler as well.

**Comment by Saklad5** [comment]
That should use the normal behavior for a direct-fire ballistic weapon. No special case there.

**Comment by Sleet01** [comment]
> That should use the normal behavior for a direct-fire ballistic weapon. No special case there.

Is this in regards to the RAC/5, or the Homing AIV?  If the latter, we have a ruling that states Glancing Blow should reduce damage from the direct hit portion of the HAIV (but not the AE damage to other occupants of the hex, obviously).

**Comment by Saklad5** [comment]
Yes, because the impact of a Homing Arrow is resolved as a direct-fire ballistic weapon, but the AoE damage is not.

**Comment by Sleet01** [comment]
> Yes, because the impact of a Homing Arrow is resolved as a direct-fire ballistic weapon, but the AoE damage is not.

Ah, I see the confusion.  You are saying we should make it use the DF, B handler, which should handle Direct Blow / Glancing Blow correctly?  Rather than that it is _currently_ working as expected and doesn't need a specific handler?

Unfortunately that won't be possible without completely rewriting all of the Arrow IV code, and probably some other stuff as well.  Hence my suggestion to fix the Arrow IV case with its own handler.

**Comment by Saklad5** [comment]
I'm proposing an eventual refactor to that end, yes. I'm a bit scared of touching that myself, though, so I can't complain about another patch job.

**Comment by Sleet01** [comment]
> I'm proposing an eventual refactor to that end, yes. I'm a bit scared of touching that myself, though, so I can't complain about another patch job.

The problem is that all of our attack handling is set up with inheritance, but we probably want interfaces so we can mix and match them.  Since Homing AIV _is_ Artillery, but _treated_ as DF, B on successful guided hit, our current system won't give us all of the correct behavior with inheritance alone.

**Comment by Saklad5** [comment]
This is the sort of thing that leads me to heavily favor protocol-oriented programming over object-oriented programming when I have the option. I'm pretty sure Java interfaces didn't exist when this code was first made, though.
